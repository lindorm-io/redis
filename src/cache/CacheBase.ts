import Joi from "@hapi/joi";
import { IEntity } from "@lindorm-io/core";
import { Logger } from "@lindorm-io/winston";
import { RedisClient, RedisInMemoryClient } from "../class";

export interface ICache<Entity> {
  create(entity: Entity): Promise<Entity>;
  find(id: string): Promise<Entity>;
  findAll(): Promise<Array<Entity>>;
  remove(entity: Entity): Promise<void>;
}

export interface ICacheOptions {
  expiresInSeconds?: number;
  logger: Logger;
  client: RedisClient | RedisInMemoryClient;
}

export interface ICacheBaseOptions extends ICacheOptions {
  entityName: string;
  schema: Joi.Schema;
}

export abstract class CacheBase<Entity extends IEntity> implements ICache<Entity> {
  private client: RedisClient | RedisInMemoryClient;
  private expiresInSeconds: number;
  private prefix: string;
  private schema: Joi.Schema;
  protected logger: Logger;

  protected constructor(options: ICacheBaseOptions) {
    this.client = options.client;
    this.expiresInSeconds = options.expiresInSeconds || null;
    this.prefix = options.entityName;
    this.schema = options.schema;

    this.logger = options.logger.createChildLogger(["redis", "cache", options.entityName]);
  }

  protected abstract createEntity(data: IEntity): Entity;

  protected abstract getEntityJSON(entity: Entity): IEntity;

  async create(entity: Entity): Promise<Entity> {
    const start = Date.now();
    const json = this.getEntityJSON(entity);
    const key = `${this.prefix}::${json.id}`;

    await this.schema.validateAsync(json);
    await this.client.set(key, json, this.expiresInSeconds);

    this.logger.debug("set", {
      payload: Object.keys(json),
      time: Date.now() - start,
    });

    return entity;
  }

  async find(id: string): Promise<Entity> {
    const start = Date.now();

    const key = `${this.prefix}::${id}`;
    const result: unknown = await this.client.get(key);
    const data = result as IEntity;

    this.logger.debug("get", {
      id,
      result: { success: !!result },
      time: Date.now() - start,
    });

    if (!result) {
      throw new Error("Not found");
    }

    return this.createEntity(data);
  }

  async findAll(): Promise<Array<Entity>> {
    const start = Date.now();

    const pattern = `${this.prefix}::*`;
    const result: any = await this.client.getAll(pattern);
    const data: Array<Entity> = [];

    for (const object of Object.values(result)) {
      data.push(this.createEntity(object as IEntity));
    }

    this.logger.debug("getAll", {
      pattern,
      result: { success: !!result },
      time: Date.now() - start,
    });

    if (!result) {
      throw new Error("Not found");
    }

    return data;
  }

  async remove(entity: Entity): Promise<void> {
    const start = Date.now();

    const { id } = this.getEntityJSON(entity);
    const key = `${this.prefix}::${id}`;

    await this.client.del(key);

    this.logger.debug("del", {
      filter: Object.keys({ id }),
      result: { success: true },
      time: Date.now() - start,
    });
  }
}
