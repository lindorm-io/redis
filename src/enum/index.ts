export * from "./redis-connection-type";
