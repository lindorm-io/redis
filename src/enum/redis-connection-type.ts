export enum RedisConnectionType {
  MEMORY = "MEMORY",
  CACHE = "CACHE",
}
