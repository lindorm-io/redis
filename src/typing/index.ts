export * from "./cache";
export * from "./redis-client";
export * from "./redis-connection";
