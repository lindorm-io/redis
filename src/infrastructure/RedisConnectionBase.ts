import { ClientOpts } from "redis";
import { RedisClient } from "../class";
import { IRedisConnection, RedisConnectionOptions } from "../typing";

export abstract class RedisConnectionBase implements IRedisConnection {
  protected readonly clientOptions: ClientOpts;
  protected _client: RedisClient | undefined;

  protected constructor(options: RedisConnectionOptions) {
    const { inMemoryCache, type, ...clientOptions } = options;

    this.clientOptions = clientOptions;
  }

  public abstract connect(): Promise<void>;

  public abstract disconnect(): Promise<void>;

  public isConnected(): boolean {
    if (!this._client) {
      return false;
    }
    return this._client.isConnected();
  }

  public client(): RedisClient {
    if (!this._client) {
      throw new Error("Client could not be found. Call connect() first.");
    }
    return this._client;
  }
}
