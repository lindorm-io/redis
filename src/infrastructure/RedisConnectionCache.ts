import { IRedisConnection, IRedisConnectionOptions } from "../typing";
import { RedisClient } from "../class";
import { RedisConnectionBase } from "./RedisConnectionBase";

export class RedisConnectionCache extends RedisConnectionBase implements IRedisConnection {
  public constructor(options: IRedisConnectionOptions) {
    super(options);
  }

  public async connect(): Promise<void> {
    this._client = new RedisClient(this.clientOptions);
    await this._client.connect();
  }

  public async disconnect(): Promise<void> {
    if (!this._client) return;
    await this._client.quit();
  }
}
