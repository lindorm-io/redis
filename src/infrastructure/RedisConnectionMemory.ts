import { IRedisConnection, IRedisConnectionBaseOptions } from "../typing";
import { RedisConnectionBase } from "./RedisConnectionBase";
import { RedisInMemoryClient } from "../class";

export class RedisConnectionMemory extends RedisConnectionBase implements IRedisConnection {
  public client: RedisInMemoryClient;
  public inMemoryCache: Record<string, any>;

  constructor(options: IRedisConnectionBaseOptions) {
    super(options);
    this.inMemoryCache = options.inMemoryCache;
  }

  public async connect(): Promise<void> {
    this.client = new RedisInMemoryClient(this.inMemoryCache);
    await this.client.connect();
  }

  public async disconnect(): Promise<void> {
    await this.client.quit();
  }
}
