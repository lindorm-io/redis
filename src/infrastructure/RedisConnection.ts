import { IRedisConnection, IRedisConnectionOptions, TRedisClient } from "../typing";
import { RedisConnectionBase } from "./RedisConnectionBase";
import { RedisConnectionCache } from "./RedisConnectionCache";
import { RedisConnectionMemory } from "./RedisConnectionMemory";
import { RedisConnectionType } from "../enum";

export class RedisConnection implements IRedisConnection {
  private connection: RedisConnectionBase;

  constructor(options: IRedisConnectionOptions) {
    switch (options.type) {
      case RedisConnectionType.CACHE:
        this.connection = new RedisConnectionCache(options);
        break;

      case RedisConnectionType.MEMORY:
        this.connection = new RedisConnectionMemory(options);
        break;

      default:
        throw new Error(`Unknown connection type: [ ${options.type} ]`);
    }
  }

  public async connect(): Promise<void> {
    return this.connection.connect();
  }

  public async disconnect(): Promise<void> {
    return this.connection.disconnect();
  }

  public getClient(): TRedisClient {
    return this.connection.getClient();
  }
}
