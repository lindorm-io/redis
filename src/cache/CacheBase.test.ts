import Joi from "@hapi/joi";
import MockDate from "mockdate";
import { CacheBase, ICache, ICacheOptions } from "./CacheBase";
import { EntityBase, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";
import { Logger, LogLevel } from "@lindorm-io/winston";
import { RedisConnection } from "../infrastructure";
import { RedisConnectionType } from "../enum";
import { TRedisClient } from "../typing";

jest.mock("uuid", () => ({
  v4: () => "e397bc49-849e-4df6-a536-7b9fa3574ace",
}));

MockDate.set("2020-01-01 08:00:00.000");

interface IMockEntity extends IEntity {
  name: string;
}

interface IMockEntityOptions extends IEntityBaseOptions {
  name: string;
}

class MockEntity extends EntityBase implements IMockEntity {
  public name: string;

  constructor(options: IMockEntityOptions) {
    super(options);
    this.name = options.name;
  }
}

interface IMockCache extends ICache<MockEntity> {
  create(entity: MockEntity): Promise<MockEntity>;
  find(id: string): Promise<MockEntity>;
  remove(entity: MockEntity): Promise<void>;
}

class MockCache extends CacheBase<MockEntity> implements IMockCache {
  constructor(options: ICacheOptions) {
    super({
      ...options,
      entityName: "mock",
      schema: Joi.object(),
    });
  }

  protected createEntity(data: IMockEntity): MockEntity {
    return new MockEntity(data);
  }

  protected getEntityJSON(entity: MockEntity): IMockEntity {
    return {
      id: entity.id,
      created: entity.created,
      updated: entity.updated,
      version: entity.version,

      name: entity.name,
    };
  }

  protected getEntityKey(entity: MockEntity): string {
    return entity.id;
  }
}

const logger = new Logger({
  packageName: "n",
  packageVersion: "v",
});
logger.addConsole(LogLevel.ERROR);

describe("CacheBase", () => {
  let inMemoryCache: TObject<any>;
  let redis: RedisConnection;
  let client: TRedisClient;
  let cache: MockCache;
  let entity: MockEntity;

  beforeEach(async () => {
    inMemoryCache = {};

    redis = new RedisConnection({
      type: RedisConnectionType.MEMORY,
      port: 1,
      inMemoryCache,
    });

    await redis.connect();
    client = redis.getClient();

    cache = new MockCache({
      logger,
      client,
    });
    entity = new MockEntity({
      name: "name",
    });
  });

  afterEach(async () => {
    await client.quit();
    jest.clearAllMocks();
  });

  test("should create entity", async () => {
    await expect(cache.create(entity)).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should create entity with expiry", async () => {
    cache = new MockCache({
      client,
      expiresInSeconds: 100,
      logger,
    });

    await expect(cache.create(entity)).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should find entity", async () => {
    await cache.create(entity);

    await expect(cache.find(entity.id)).resolves.toMatchSnapshot();
  });

  test("should find all entities", async () => {
    await cache.create(
      new MockEntity({
        id: "e397bc49-849e-4df6-a536-7b9fa3574ace",
        name: "one",
      }),
    );
    await cache.create(
      new MockEntity({
        id: "fa354ace-4df6-849e-a536-e397b7c497b9",
        name: "two",
      }),
    );

    await expect(cache.findAll()).resolves.toMatchSnapshot();
  });

  test("should remove entity", async () => {
    await cache.create(entity);

    await expect(cache.remove(entity)).resolves.toBe(undefined);

    expect(inMemoryCache).toMatchSnapshot();
  });
});
