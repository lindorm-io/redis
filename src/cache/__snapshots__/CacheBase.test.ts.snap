// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheBase should create entity 1`] = `
MockEntity {
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
  "name": "name",
}
`;

exports[`CacheBase should create entity 2`] = `
Object {
  "blob": "{\\"json\\":{\\"id\\":\\"e397bc49-849e-4df6-a536-7b9fa3574ace\\",\\"created\\":\\"2020-01-01T07:00:00.000Z\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\",\\"version\\":\\"0\\",\\"name\\":\\"name\\"},\\"meta\\":{\\"id\\":\\"string\\",\\"created\\":\\"date\\",\\"updated\\":\\"date\\",\\"version\\":\\"number\\",\\"name\\":\\"string\\"}}",
  "expiresInSeconds": null,
}
`;

exports[`CacheBase should create entity with expiry 1`] = `
MockEntity {
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
  "name": "name",
}
`;

exports[`CacheBase should create entity with expiry 2`] = `
Object {
  "blob": "{\\"json\\":{\\"id\\":\\"e397bc49-849e-4df6-a536-7b9fa3574ace\\",\\"created\\":\\"2020-01-01T07:00:00.000Z\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\",\\"version\\":\\"0\\",\\"name\\":\\"name\\"},\\"meta\\":{\\"id\\":\\"string\\",\\"created\\":\\"date\\",\\"updated\\":\\"date\\",\\"version\\":\\"number\\",\\"name\\":\\"string\\"}}",
  "expiresInSeconds": 100,
}
`;

exports[`CacheBase should find all entities 1`] = `
Array [
  MockEntity {
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [],
    "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
    "name": "one",
  },
  MockEntity {
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [],
    "_id": "fa354ace-4df6-849e-a536-e397b7c497b9",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
    "name": "two",
  },
]
`;

exports[`CacheBase should find entity 1`] = `
MockEntity {
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
  "name": "name",
}
`;

exports[`CacheBase should remove entity 1`] = `undefined`;
